File: App/Sources/Environment/SupabaseSettings.swift
import Foundation
import TeamTasksSupabase

/// Reads the Supabase project settings that project.yml writes into Info.plist from the build settings
/// (Config/Base.xcconfig, overridden by the generated Config/Secrets.xcconfig).
enum SupabaseSettings {
    static let schemeKey = "SupabaseScheme"
    static let hostKey = "SupabaseHost"
    static let publishableKeyKey = "SupabasePublishableKey"

    /// Values of Config/Secrets.example.xcconfig: copying the example without editing it is not a configuration.
    private static let exampleHost = "your-project-ref.supabase.co"
    private static let exampleKey = "sb_publishable_xxxxxxxxxxxxxxxx"

    /// The configuration of this bundle's Info.plist, or every problem found.
    static func configuration(from bundle: Bundle) -> Result<SupabaseConfiguration, ConfigurationIssue> {
        configuration { key in bundle.object(forInfoDictionaryKey: key) }
    }

    /// The configuration from Info.plist-like values, or every problem found (all of them at once, so that a
    /// single rebuild fixes everything).
    static func configuration(lookup: (String) -> Any?) -> Result<SupabaseConfiguration, ConfigurationIssue> {
        var problems: [ConfigurationIssue.Problem] = []

        let scheme = (usableValue(lookup(schemeKey)) ?? "https").lowercased()
        if scheme != "https" && scheme != "http" {
            problems.append(.invalidScheme(scheme))
        }

        var url: URL?
        switch usableValue(lookup(hostKey)) {
        case nil:
            problems.append(.missingHost)
        case let host? where host == exampleHost:
            problems.append(.exampleHost)
        case let host? where host.hasSuffix(":") || host.contains("://"):
            // `SUPABASE_HOST = https://…` in an .xcconfig file: "//" starts a comment, only "https:" is left.
            problems.append(.hostWithScheme(host))
        case let host?:
            let trimmedHost = host.hasSuffix("/") ? String(host.dropLast()) : host
            if let candidate = URL(string: "\(scheme)://\(trimmedHost)"), let parsedHost = candidate.host(), !parsedHost.isEmpty {
                url = candidate
            } else {
                problems.append(.invalidHost(host))
            }
        }

        let key = usableValue(lookup(publishableKeyKey))
        switch key {
        case nil:
            problems.append(.missingKey)
        case let key? where key == exampleKey:
            problems.append(.exampleKey)
        default:
            break
        }

        guard problems.isEmpty, let url, let key else {
            return .failure(ConfigurationIssue(problems: problems))
        }
        return .success(SupabaseConfiguration(url: url, publishableKey: key))
    }

    /// A trimmed Info.plist string; nil when missing, empty, or an unexpanded `$(BUILD_SETTING)`.
    private static func usableValue(_ raw: Any?) -> String? {
        guard let raw = raw as? String else { return nil }
        let trimmed = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty, !trimmed.hasPrefix("$(") else { return nil }
        return trimmed
    }
}

/// Why the app cannot reach its backend: shown by the « Configuration manquante » screen.
struct ConfigurationIssue: Error, Sendable, Equatable {
    enum Problem: Sendable, Equatable {
        case missingHost
        case exampleHost
        case hostWithScheme(String)
        case invalidHost(String)
        case invalidScheme(String)
        case missingKey
        case exampleKey

        /// French explanation of the problem.
        var message: String {
            switch self {
            case .missingHost:
                "SUPABASE_HOST est vide : indiquez l’hôte de votre projet, par exemple « abcdefgh.supabase.co »."
            case .exampleHost:
                "SUPABASE_HOST contient encore la valeur d’exemple « your-project-ref.supabase.co »."
            case let .hostWithScheme(host):
                "SUPABASE_HOST vaut « \(host) » : indiquez seulement l’hôte, sans « https:// » (dans un fichier .xcconfig, « // » commence un commentaire)."
            case let .invalidHost(host):
                "SUPABASE_HOST « \(host) » n’est pas un nom d’hôte valide."
            case let .invalidScheme(scheme):
                "SUPABASE_SCHEME vaut « \(scheme) » : utilisez « https » (ou « http » pour un Supabase local)."
            case .missingKey:
                "SUPABASE_PUBLISHABLE_KEY est vide : copiez la clé publique « sb_publishable_… » de votre projet."
            case .exampleKey:
                "SUPABASE_PUBLISHABLE_KEY contient encore la valeur d’exemple."
            }
        }
    }

    var problems: [Problem]
}

File: android/app/src/main/kotlin/io/github/notkanaa/equipe/config/SupabaseSettings.kt
package io.github.notkanaa.equipe.config

import io.github.notkanaa.equipe.BuildConfig
import io.github.notkanaa.equipe.supabase.SupabaseBackend
import io.github.notkanaa.equipe.supabase.SupabaseConfiguration
import java.net.URI
import java.net.URISyntaxException
import java.util.Locale

/** Why the app cannot reach its backend: every problem found, shown by the « Configuration manquante » screen. */
data class ConfigurationIssue(val problems: List<Problem>) {
    sealed interface Problem {
        /** French explanation of the problem. */
        val message: String

        data object MissingUrl : Problem {
            override val message: String =
                "SUPABASE_URL est vide : indiquez l’adresse de votre projet, par exemple " +
                    "« https://abcdefgh.supabase.co »."
        }

        data object ExampleUrl : Problem {
            override val message: String =
                "SUPABASE_URL contient encore la valeur d’exemple « https://${SupabaseSettings.EXAMPLE_HOST} »."
        }

        data class InvalidUrl(val url: String) : Problem {
            override val message: String =
                "SUPABASE_URL « $url » n’est pas une adresse valide : indiquez « https:// » " +
                    "suivi de l’hôte du projet (ou « http:// » pour un Supabase local), sans chemin."
        }

        data object MissingKey : Problem {
            override val message: String =
                "SUPABASE_PUBLISHABLE_KEY est vide : copiez la clé publique « sb_publishable_… » de " +
                    "votre projet."
        }

        data object ExampleKey : Problem {
            override val message: String = "SUPABASE_PUBLISHABLE_KEY contient encore la valeur d’exemple."
        }

        /** A secret (`sb_secret_…`) or `service_role` key instead of the publishable one. */
        data object SecretKey : Problem {
            override val message: String =
                "SUPABASE_PUBLISHABLE_KEY contient une clé secrète, qui donnerait accès à toutes les données à " +
                    "quiconque l’extrait de l’app : utilisez la clé publique « sb_publishable_… » de " +
                    "votre projet, et régénérez la clé secrète si elle a été partagée."
        }
    }
}

/**
 * The Supabase project the app was built for: BuildConfig.SUPABASE_URL and SUPABASE_PUBLISHABLE_KEY, set at build time
 * from the Gradle properties `equipe.supabaseUrl` / `equipe.supabaseKey` or the environment variables SUPABASE_URL /
 * SUPABASE_PUBLISHABLE_KEY (port of the iOS `SupabaseSettings`).
 */
object SupabaseSettings {
    /** Values of the iOS Config/Secrets.example.xcconfig: copying an example is not a configuration. */
    const val EXAMPLE_HOST: String = "your-project-ref.supabase.co"
    const val EXAMPLE_KEY: String = "sb_publishable_xxxxxxxxxxxxxxxx"

    sealed interface Resolution {
        data class Configured(val configuration: SupabaseConfiguration) : Resolution

        data class Misconfigured(val issue: ConfigurationIssue) : Resolution
    }

    /**
     * The configuration, or every problem found (all at once, so that a single rebuild fixes everything).
     *
     * @param keyWithheld the build refused to embed the key because it is a secret key.
     */
    fun resolve(
        url: String = BuildConfig.SUPABASE_URL,
        publishableKey: String = BuildConfig.SUPABASE_PUBLISHABLE_KEY,
        keyWithheld: Boolean = BuildConfig.SUPABASE_KEY_IS_SECRET,
    ): Resolution {
        val problems = ArrayList<ConfigurationIssue.Problem>()

        val trimmedUrl = url.trim()
        var projectUrl: String? = null
        when {
            trimmedUrl.isEmpty() -> problems.add(ConfigurationIssue.Problem.MissingUrl)
            trimmedUrl.lowercase(Locale.ROOT).contains(EXAMPLE_HOST) ->
                problems.add(ConfigurationIssue.Problem.ExampleUrl)
            else -> {
                projectUrl = projectUrl(trimmedUrl)
                if (projectUrl == null) problems.add(ConfigurationIssue.Problem.InvalidUrl(trimmedUrl))
            }
        }

        val key = publishableKey.trim()
        when {
            keyWithheld -> problems.add(ConfigurationIssue.Problem.SecretKey)
            key.isEmpty() -> problems.add(ConfigurationIssue.Problem.MissingKey)
            key == EXAMPLE_KEY -> problems.add(ConfigurationIssue.Problem.ExampleKey)
            // Shipped in the app, a secret key would give everyone who extracts it access to every row (it bypasses
            // RLS): refuse to run with it rather than use it.
            SupabaseBackend.isSecretKey(key) -> problems.add(ConfigurationIssue.Problem.SecretKey)
        }

        if (problems.isNotEmpty() || projectUrl == null) {
            return Resolution.Misconfigured(ConfigurationIssue(problems))
        }
        return Resolution.Configured(SupabaseConfiguration(url = projectUrl, publishableKey = key))
    }

    /**
     * `scheme://host[:port]` of an `http(s)` URL with a host and no path, query, fragment or user info (trailing
     * slashes ignored), else null: the URLs `SupabaseBackend.makeServices` accepts (it throws for the others).
     */
    fun projectUrl(url: String): String? {
        val trimmed = url.trim().trimEnd('/')
        val uri = try {
            URI(trimmed)
        } catch (error: URISyntaxException) {
            return null
        }
        val scheme = uri.scheme ?: return null
        val valid = (scheme == "https" || scheme == "http") && !uri.host.isNullOrEmpty() &&
            uri.rawPath.isNullOrEmpty() && uri.rawQuery == null && uri.rawFragment == null && uri.rawUserInfo == null
        return if (valid) trimmed else null
    }
}
